WT-11709 API to retrieve timestamp of a checkpoint cursor

marcbutler opened the pull request:
This also encompasses creating a more general interface for querying cursors for detailed information.

marcbutler commented:
Hi @ddanderson minimal prototypes in `wiredtiger.in` for you to review before any development begins. Thanks.

ddanderson reviewed (approved):
@marcbutler that looks just right to me.

ddanderson reviewed (commented):
@marcbutler This is headed in the right direction - there's a type mismatch issue with different WT_CURSOR_* flavors that needs to be addressed first.  Happy to chat about it.

marcbutler commented:
@ddanderson Thanks. I understood the subclassing but I did not understand that a checkpoint is a "file" so-to-speak.

I fell back on return `ENOTSUP` for all other cursor types as `ENODATA` alludes to a transient error.

Q: Should I be return all 3 timestamps?

marcbutler commented on src/cursor/cur_ds.c:
```diff
@@ -453,7 +453,9 @@ __wt_curds_open(WT_SESSION_IMPL *session, const char *uri, WT_CURSOR *owner, con
       __wt_cursor_checkpoint_id,                      /* checkpoint ID */
-      __curds_close);                                 /* close */
+      __curds_close,                                  /* close */
+      __wt_cursor_get_details                         /* get_details */
```
Thanks I believe I have done this correctly, but thought ENOTSUP was better for cursor types that do not yet support the call.

ddanderson replied:
I agree that ENOTSUP is the right return

But I haven't seen the corrected code, have you pushed it?  I think the case that would fail would be to create some kind of "other" cursor, like a backup cursor or metadata cursor and then call cursor.get_details on it.  The place where the code falls down is:
```
    cbt = (WT_CURSOR_BTREE *)cursor;
    if (cbt->checkpoint_txn) {
      ...
```
because in those other cursor types, `cursor` will not be a `(WT_CURSOR_BTREE *)`.  Perhaps I'm missing something.

You'll want to have a test case like that.

marcbutler commented on src/cursor/cur_file.c:
```diff
@@ -106,6 +106,26 @@ __wt_cursor_checkpoint_id(WT_CURSOR *cursor)
+ *     Query detailed information using this cursor instance.
+ */
+int
+__wt_cursor_get_details(WT_CURSOR *cursor, WT_CURSOR_DETAILS *detailsp, const char *config)
```
👍

marcbutler commented:
So this is running a foul of the SWIG code generation, but I don't quite have a handle on as to why.

ddanderson commented:
@marcbutler Re SWIG: you might look at lang/python/wiredtiger.i , that's the SWIG input file.  It's pretty hard to fathom, but I think Monica has been in there, might be able to help.  The API you're adding is passing a pointer to a struct that is being filled in and "returned", you might want to look at the WT_MODIFY (or __wt_modify) struct and how it's handled in SWIG.  That one is a little harder than yours, because it is passing an array of WT_MODIFY structs as input for one function and output for another function.  You only need to pass a single struct and have the new API fill it in.  Another thing to know about SWIG is it  generates its output (C and python wrapper files) in <your_build_directory>/lang/python/wiredtiger/

If this is too hard to resolve, could be possible to spin the Python off into a separate ticket, and insert
```
#if !defined(SWIG)
```
around the function definition.

Whether or not you have a Python interface, there should be tests for the new API.

ddanderson reviewed (commented):
I think we still need to fix the typing issue.  We can meet if that helps.

marcbutler commented:
I would have liked a cleaner definition of `__wt_cursor_details` but at some point struggling with SWIG just did not seem to be adding value.

I have not added a `cppsuite` test as testing through the python interface takes care of everything.

I am wondering if the Python `get_details` method should wrap the underlying C call, check the return code and throw if it is non-zero. This would disambiguate the interface which only returns two values if successful.

ddanderson commented on test/suite/test_cursor_details01.py:
```diff
@@ -0,0 +1,34 @@
+        self.session.create(uri, "key_format=S,value_format=S")
+        self.session.checkpoint()
+
+        # Query checkpoint cursor details. This is the first and currently
+        # only cursor to support this API all other should return ENOTSUP.
+        cursor = self.session.open_cursor(uri, None, f"checkpoint={internal_checkpoint_name}")
+        ret, details = cursor.get_details(None)
+        
```
It would be nice to check that we're doing the two things we're expected to do.  That is:

- add some data, advance stable timestamp, do a checkpoint, open a cursor on that checkpoint and validate the timestamp.
- do a checkpoint, open a cursor on most recent checkpoint, open a second cursor on recent checkpoint, validate they have the same id, do another checkpoint and open a third cursor, and validate they have diff ids.

marcbutler commented:
@ddanderson @agorrod

Don I hope this commit addresses your concerns.

The testing in Python has been expanded. I can see no value in adding a cppsuite test, as the C interface is implicitly tested via Python.

I've added a generated wrapper so the Python `get_details()` will throw an exception if called a cursor type that does not support it: which gives a much cleaner Python API.

ddanderson reviewed (approved):
LGTM!

marcbutler commented:
@agorrod Has concerns and has taken this ticket on.
